core/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/SignificantTextAggregator.java

```diff
@@ -0,0 +1,253 @@
+            
+            @Override
+            public void collect(int doc, long bucket) throws IOException {
+                loadFromSource(doc, bucket, fieldName);
+                numCollectedDocs++;
+                if (dupSequenceSpotter != null) {
+                    dupSequenceSpotter.startNewSequence();
+                }
+            }
+            
+            private void processTokenStream(int doc, long bucket, TokenStream ts, String fieldText) throws IOException{
+                if (dupSequenceSpotter != null) {
+                    ts = new DeDuplicatingTokenFilter(ts, dupSequenceSpotter);
+                }
+                CharTermAttribute termAtt = ts.addAttribute(CharTermAttribute.class);
+                ts.reset();
+                try {
+                    //Assume tokens will average 5 bytes in length to size number of tokens
+                    BytesRefHash inDocTerms = new BytesRefHash(1+(fieldText.length()/5), context.bigArrays());
+                    
+                    try{
+                        while (ts.incrementToken()) {
+                            if (dupSequenceSpotter != null) {
+                                long newTrieSize = dupSequenceSpotter.getEstimatedSizeInBytes();
+                                long growth = newTrieSize - lastTrieSize;
+                                // Only update the circuitbreaker after
+                                if (growth > MEMORY_GROWTH_REPORTING_INTERVAL_BYTES) {
+                                    addRequestCircuitBreakerBytes(growth);
+                                    lastTrieSize = newTrieSize;
+                                }
+                            }
+                            previous.clear();
+                            previous.copyChars(termAtt);
+                            BytesRef bytes = previous.get();
+                            if (inDocTerms.add(bytes) >= 0) {
+                                if (includeExclude == null || includeExclude.accept(bytes)) {
+                                    long bucketOrdinal = bucketOrds.add(bytes);
+                                    if (bucketOrdinal < 0) { // already seen
+                                        bucketOrdinal = -1 - bucketOrdinal;
+                                        collectExistingBucket(sub, doc, bucketOrdinal);
+                                    } else {
+                                        collectBucket(sub, doc, bucketOrdinal);
+                                    }
+                                }
+                            }
+                        }
+                    }
+                    finally{
+                        Releasables.close(inDocTerms);
+                    }
+                }
+                finally{
+                    ts.close();
+                }
+            }
+            
+            private void loadFromSource(int doc, long bucket, String fieldName) throws IOException {
```

Comment:
Would `collectFromSource()` be a better name for this since the outcome of this method is that the document is collected and placed in a bucket?